gx/gx-core/src/main/java/io/graphenee/core/impl/GxMailServiceImpl.java

```diff
@@ -15,39 +15,77 @@
  *******************************************************************************/
 package io.graphenee.core.impl;
 
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
 import java.util.Date;
+import java.util.List;
 
-import org.springframework.mail.SimpleMailMessage;
+import javax.mail.MessagingException;
+import javax.mail.internet.MimeMessage;
+
+import org.springframework.core.io.ByteArrayResource;
+import org.springframework.mail.MailSendException;
 import org.springframework.mail.javamail.JavaMailSender;
+import org.springframework.mail.javamail.MimeMessageHelper;
 import org.springframework.util.StringUtils;
 
+import io.graphenee.core.exception.SendMailFailedException;
+import io.graphenee.core.model.bean.GxMailAttachmentBean;
+import io.graphenee.core.storage.FileStorage;
+import io.graphenee.core.storage.ResolveFailedException;
+
 public class GxMailServiceImpl implements io.graphenee.core.api.GxMailService {
 
 	private JavaMailSender javaMailSender;
 
+	private FileStorage fileStorage;
+
 	@Override
-	public void sendEmail(String subject, String content, String senderEmail, String recipientEmail, String ccEmailList, String bccEmailList) {
-		SimpleMailMessage message = new SimpleMailMessage();
-		message.setSubject(subject);
-		message.setText(content);
-		message.setTo(recipientEmail);
-		message.setFrom(senderEmail);
-		message.setSentDate(new Date());
-		if (!StringUtils.isEmpty(bccEmailList))
-			message.setBcc(bccEmailList);
-		if (!StringUtils.isEmpty(ccEmailList))
-			message.setCc(ccEmailList);
-		getJavaMailSender().send(message);
+	public void sendEmail(String subject, String content, String senderEmail, String recipientEmail, String ccEmailList, String bccEmailList,
+			List<GxMailAttachmentBean> attachmentBeans) throws Exception {
+		try {
+			MimeMessage message = javaMailSender.createMimeMessage();
+			MimeMessageHelper mimeMessage = new MimeMessageHelper(message, true);
+			mimeMessage.setSubject(subject);
+			mimeMessage.setText(content);
+			mimeMessage.setTo(recipientEmail);
+			mimeMessage.setFrom(senderEmail);
+			mimeMessage.setSentDate(new Date());
+			if (!StringUtils.isEmpty(bccEmailList))
+				mimeMessage.setBcc(bccEmailList);
+			if (!StringUtils.isEmpty(ccEmailList))
+				mimeMessage.setCc(ccEmailList);
+			if (attachmentBeans != null)
+				for (GxMailAttachmentBean attachmentBean : attachmentBeans) {
+					String resourcePath = getFileStorage().resourcePath(attachmentBean.getAttachmentFolder(), attachmentBean.getAttachmentFilePath());
+					InputStream inputStream = getFileStorage().resolve(resourcePath);
+					mimeMessage.addAttachment(attachmentBean.getAttachmentName(), new ByteArrayResource(org.apache.poi.util.IOUtils.toByteArray(inputStream)));
+				}
+			javaMailSender.send(message);
+		} catch (ResolveFailedException e) {
+			e.printStackTrace();
+		} catch (IOException e) {
+			e.printStackTrace();
+		} catch (Exception e) {
```

Comment:
This catch clause will not catch ResolveFailedException and IOException since you already caught them above but eaten them. You must have one catch clause with catch(Exception e) and throw a new SendMailFailedException wrapping that exception. Please ask Hafiz for guidance.